projects/bifrost/index.js

```diff
@@ -0,0 +1,109 @@
+const BigNumber = require("bignumber.js");
+const web3 = require("../config/web3");
+const abi = require("./abi.json");
+
+// node test.js projects/bifrost/index.js
+function formatTokenAmount(amount, tokenSymbol) {
+  let decimals = 12;
+
+  switch (tokenSymbol) {
+    case "DOT":
+    case "RMRK":
+      decimals = 10;
+      break;
+
+    case "KSM":
+    case "KAR":
+    case "KUSD":
+      decimals = 12;
+      break;
+    case "ETH":
+    case "ZLK":
+      decimals = 18;
+      break;
+  }
+
+  return Number(amount / Number(10 ** decimals));
+}
+
+function formatToken(token) {
+  switch (token) {
+    case `{"Token":"RMRK"}`:
+      return "RMRK";
+    case `{"Token":"KSM"}`:
+      return "KSM";
+    case `{"VSToken":"KSM"}`:
+      return "KSM";
+    case `{"Token":"KAR"}`:
+      return "KAR";
+    case `{"Token":"ZLK"}`:
+      return "ZLK";
+    case `{"Stable":"KUSD"}`:
+      return "KUSD";
+    default :
+      return null;
+  }
+}
+
+const tokenToCoingecko = {
+  DOT: "polkadot",
+  KSM: "kusama",
+  KAR: "karura",
+  KUSD: "tether",
+  ZLK: "zenlink-network-token",
+  RMRK: "rmrk",
+  ETH: "ethereum"
+};
+
+async function tvl() {
+  const provider = new WsProvider("wss://bifrost-rpc.liebi.com/ws");
+  const api = await ApiPromise.create(options({ provider }));
+
+  // Get Salp tvl
+  const vsKSM = (await api.query.tokens.totalIssuance({ "vsToken": "KSM" })).toString();
+  const vsDOT = (await api.query.tokens.totalIssuance({ "vsToken": "DOT" })).toString();
+
+  const totalLiquidity = {};
+  totalLiquidity.KSM = vsKSM;
+  totalLiquidity.DOT = vsDOT;
+
+  // Get swap tvl
+  const pools = await api.query.zenlinkProtocol.pairStatuses.entries();
+  await Promise.all(pools.map(async (pool) => {
+    if (pool[1].toHuman()?.Trading?.pairAccount) {
+      const poolAccountTokens = await api.query.tokens.accounts.entries(pool[1].toHuman()?.Trading?.pairAccount);
+      poolAccountTokens.map(item => {
+        const currentToken = formatToken(JSON.stringify(item[0].toHuman()[1]));
+        if (currentToken) {
+          if (totalLiquidity[currentToken]) {
+            totalLiquidity[currentToken] = new BigNumber(totalLiquidity[currentToken]).plus(item[1].toJSON().free).toString();
+          } else {
+            totalLiquidity[currentToken] = new BigNumber(item[1].toJSON().free).toString();
+          }
+        }
+      });
+    }
+  }));
+
+  // Get vETH tvl
+  const vethContract = new web3.eth.Contract(abi, "0xc3d088842dcf02c13699f936bb83dfbbc6f721ab");
+  const totalSupply = await vethContract.methods.totalSupply().call();
+  totalLiquidity["ETH"] = totalSupply;
+
+  const totalLiquidityFormatted = {};
+  for (const key in totalLiquidity) {
+    totalLiquidityFormatted[tokenToCoingecko[key]] = formatTokenAmount(
+      totalLiquidity[key],
+      key
+    );
+  }
+
+  return totalLiquidityFormatted;
+}
+
+module.exports = {
+  methodology: "Counts all liquidity on DEX pools. KUSD is counted as USDT",
```

Comment:
can you add a key `timetravel: false,` here

This way, we would know that we cant refill old data for this adapter

Reply:
timetravel is added